html/inc/delete_account.inc

```diff
@@ -0,0 +1,187 @@
+//  - clear domain_name, last_ip_addr
+//
+function obfuscate_account($user) {
+    insert_deleted_records($user);
+    $x = "deleted_".time()."_".random_string();
+    $retval = $user->update("email_addr='$x', authenticator='$x', name='deleted', country='', postal_code='', has_profile=0");
+    if (!$retval) return false;
+    user_quit_team($user);
+    forum_delete_user($user);
+    pm_delete_user($user);
+    anonymize_hosts($user);
+    delete_profile($user);
+    delete_friends($user);
+    return true;
+}
+
+
+// return true if the result is in progress
+function is_in_progress($res) {
+    if ($res->server_state == RESULT_SERVER_STATE_IN_PROGRESS) {
+        return true;
+    }
+    return false;
+}
+
+// returns true if the result finished successfully but is either 
+// pending validation or inconclusive
+function is_over_but_not_validated($res) {
+    if ($res->server_state == RESULT_SERVER_STATE_OVER && $res->outcome == RESULT_OUTCOME_SUCCESS &&
+            ($res->validate_state == VALIDATE_STATE_INIT || $res->validate_state == VALIDATE_STATE_INCONCLUSIVE) ) {
+        return true;
+    }    
+    return false;
+}
+
+function transition_workunit($res) {
+    $now = time();
+    BoincWorkunit::update_aux("transition_time=$now where id=$res->workunitid");
+}
+
+// This method handles dissassociating the user from their results.
+// It will cancel all in progress or returned, but not yet validated 
+// results for a user.  For other results, it will set the userid and 
+// hostid fields to 0
+function cancel_results_for_user($user) {
+    $ress = BoincResult::enum("userid = $user->id");
+    $cancel_clause="server_state=".RESULT_SERVER_STATE_OVER.", outcome=".RESULT_OUTCOME_CLIENT_DETACHED.", validate_state=".VALIDATE_STATE_INVALID;
+    $set_id_clause="hostid=0, userid=0";
+    foreach($ress as $res) {
+        if (is_in_progress($res)) {
+            $res->update($cancel_clause.", ".$set_id_clause);
+            transition_workunit($res);
+        } else if (is_over_but_not_validated($res)) {
+            $res->update($cancel_clause.", ".$set_id_clause);
+            transition_workunit($res);
+        } else {
+            $res->update($set_id_clause);
+        }
+    }
+}
+
+// This method deletes all rows from the database associated with the user
+function wipe_account($user) {
+    $db = BoincDb::get();
+
+    insert_deleted_records($user);
+
+    // delete remote submit user
+    delete_remote_submit_user($user);  // from submit_util.inc
+    
+    // remove user's team records
+    user_erase_team_owner($user);      // from team.inc
+    user_quit_team($user);        // from team.inc
+    user_erase_team_delta($user); // from team.inc
+    
+    // Items that do not have logic elsewhere
+    // and do not have objects in boinc_db.inc
+    if (!$db) die("no DB connection");
+    $db->do_query("delete from credited_job where userid = $user->id");
+    $db->do_query("delete from donation_paypal where userid = $user->id");
```

Comment:
There should be an index on `donation_paypal.userid` to make this query efficient.

Reply:
added